api/lib/infrastructure/repositories/schooling-registration-repository.js

```diff
@@ -152,29 +151,18 @@ module.exports = {
     const existingSchoolingRegistrations = await this.findByOrganizationId({ organizationId });
 
     const reconciledSchoolingRegistrationsToImport = await this._reconcileSchoolingRegistrations(schoolingRegistrationsFromFile, existingSchoolingRegistrations);
-    const [schoolingRegistrationsToUpdate, schoolingRegistrationsToCreate] = await this._getStudentsListToUpdateOrCreate(reconciledSchoolingRegistrationsToImport, existingSchoolingRegistrations);
 
     try {
-      await Promise.all([
-        bluebird.mapSeries(schoolingRegistrationsToUpdate, async (schoolingRegistrationToUpdate) => {
-          const attributesToUpdate = _.omit(schoolingRegistrationToUpdate, ['id', 'createdAt']);
-          const whereConditions = {
-            'organizationId': organizationId,
-            'nationalStudentId': schoolingRegistrationToUpdate.nationalStudentId,
-          };
-
-          await knexConn('schooling-registrations')
-            .where(whereConditions)
-            .update({
-              ...attributesToUpdate,
-              isDisabled: false,
-              updatedAt: knexConn.raw('CURRENT_TIMESTAMP'),
-            });
-        }),
-        knexConn.batchInsert('schooling-registrations', schoolingRegistrationsToCreate),
-      ]);
+      const schoolingRegistrationToSave = reconciledSchoolingRegistrationsToImport.map((schoolingRegistration) => ({
+        ..._.omit(schoolingRegistration, ['id', 'createdAt']),
+        isDisabled: false,
+      }));
+      await knexConn('schooling-registrations')
+        .insert(schoolingRegistrationToSave)
+        .onConflict(['organizationId', 'nationalStudentId'])
+        .merge();
     } catch (err) {
-      if (bookshelfUtils.isUniqConstraintViolated(err)) {
+      if (knexUtils.isUniqConstraintViolated(err) || knexUtils.sameRowInsertedSeveralTime(err)) {
```

Comment:
**suggestion** : rajouter un test sur la levée d'erreur via `knexUtils.sameRowInsertedSeveralTime`

Reply:
J'attends la réponse [ici](https://github.com/1024pix/pix/pull/3489#discussion_r709900575) pour être sûr de ce que je dois écrire dans le test.

Reply:
Alors il y a déjà un cas qui couvre l'erreur 21000. Du coup c'est la 23505 qui ne doit pas être couverte.

Reply:
En regardant la définition de la table j'ai l'impression que la 23505 ne peut pas arriver. Est ce que quelqu'un peut me le confirmer (error codes pg: https://www.postgresql.org/docs/9.2/errcodes-appendix.html) ?

Reply:
[Question/Réponses] : [https://wiki.postgresql.org/wiki/UPSERT](https://wiki.postgresql.org/wiki/UPSERT)

"Cardinality violation" errors in detail
The SQL standard requires that MERGE raise a "cardinality violation" error [21] when more than one row is affected by an SQL MERGE command. The following examples show questionable usage of the ON CONFLICT UPDATE feature, which will cause PostgreSQL to raise an analogous cardinality violation error (unless otherwise noted):

Reply:
Avec le merge apparement le unique ne peut plus apparaître ?